core/modules/qdisp/XrdSsiMocks.cc

```diff
@@ -24,104 +24,283 @@
+         if (cancel) canCount++;
+         finCount++;
+         _isFIN = true;
+         if (_active) {
+            _rrMutex.unlock();
+            } else {
+            _rrMutex.unlock();
+            delete this;
+         }
     }
-    static void finish(QueryResource& qr) {
-        LOGS_DEBUG("QueryResourceDebug::finish");
-        qr.getJobQuery()->getMarkCompleteFunc()->operator ()(true);
+
+    void Reply(RespType rType) {
+         _go.wait(true);
+
+         // We may have been cancelled before being able to reply
+         //
+         if (_isCancelled(true)) return; // we are locked now
+
+         // Do requested reply
+         //
+         switch(rType) {
+              case RESP_DATA:
+                   _ReplyData();
+                   break;
+              case RESP_ERRNR:
+                   _reqP->doNotRetry();
+              case RESP_ERROR:
+                   _ReplyError();
+                   break;
+              case RESP_STRERR:
+                   _noData = true;
+                   _reqP->doNotRetry();  // Kill retries on stream errors
+                   _ReplyStream();
+                   break;
+              default:
+                   _reqP->doNotRetry();
+                   _ReplyError("Bad mock request!", 13);
+                   break;
+         }
+         _isCancelled(false);
     }
-};
 
-util::FlagNotify<bool> XrdSsiServiceMock::_go(true);
-util::Sequential<int> XrdSsiServiceMock::_count(0);
+    bool SetBuff(XrdSsiErrInfo& eRef, char* buff, int  blen) override {
 
-/** Class to fake being a request to xrootd.
- * Fire up thread that sleeps for a bit and then indicates it was successful.
- */
-void XrdSsiServiceMock::Provision(Resource *resP, unsigned short timeOut, bool userConn){
-    LOGS_DEBUG("XrdSsiServiceMock::Provision");
-    if (resP == nullptr) {
-        LOGS(_log, LOG_LVL_ERROR, "XrdSsiServiceMock::Provision() invoked with a null Resource pointer.");
-        return;
+         // We may have been cancelled while waiting
+         //
+         if (_isCancelled(true)) return false;
+         std::thread (&Agent::_StrmResp, this, &eRef, buff, blen).detach();
+         _rrMutex.unlock();
+         return true;
     }
-    QueryResource *qr = dynamic_cast<QueryResource*>(resP);
-    if (qr == nullptr) {
-        LOGS(_log, LOG_LVL_ERROR, "XrdSsiServiceMock::Provision() unexpected resource type.");
-        return;
+
+    Agent(lsst::qserv::qdisp::QueryRequest* rP, std::string rname, int rnum) :
+         XrdSsiStream(XrdSsiStream::isPassive),
+         _reqP(rP), _rName(rname), _rNum(rnum), _noData(true),
+         _isFIN(false), _active(true)
+ {
+         // Initialize a null message we will return as a response
+         //
+         lsst::qserv::proto::ProtoHeader* ph =
+                                          new lsst::qserv::proto::ProtoHeader;
+         ph->set_protocol(2);
+         ph->set_size(0);
+         ph->set_md5(std::string("d41d8cd98f00b204e9800998ecf8427"));
```

Comment:
This has been around for a long time and I wasn't sure how best to get rid of it. I think this would be better.
1) Make copy of original string.
2) Compare hash of original and copy, they should match.
3) Slightly alter copy and make new hash. They shouldn't match.

Reply:
I just copied the code from elsewhere, so I assume this is correct. However, I couldn't get this test to work anyway but I didn't want to remove the code in case we can get it to work in the future lest we forget all the mumbo jumbo one has to go through to setup a correct protobuf.